### ISSUE TITLE
backtesting custom variables.

### ISSUE THREAD

**Issue by anakein** [freqtrade/freqtrade#6781]
Raspberry 4```

ubuntu 20.04
Python 3.9
cctx 1.76.65
freqtrade 2022.3

this strategy working fine in dry and live mode but dont work in backtesting.

the problem is in the custom variables
Do custom variables not work in backtesting?
thanks.

```

import numpy as np  
import pandas as pd  
pd.options.mode.chained_assignment = None  # default='warn'
from pandas import DataFrame
from freqtrade.strategy import IStrategy
from freqtrade.strategy import CategoricalParameter, DecimalParameter, IntParameter
import talib.abstract as ta
import freqtrade.vendor.qtpylib.indicators as qtpylib
from freqtrade.persistence import Trade
import math
import logging
from functools import reduce

from datetime import datetime, timedelta, timezone
from timeit import default_timer as timer
from datetime import timedelta

logger = logging.getLogger(__name__)

def SSLChannels_ATR(dataframe, length=7):
    """
    SSL Channels with ATR: https://www.tradingview.com/script/SKHqWzql-SSL-ATR-channel/
    Credit to @JimmyNixx for python
    """
    df = dataframe.copy()

    df['ATR'] = ta.ATR(df, timeperiod=14)
    df['smaHigh'] = df['high'].rolling(length).mean() + df['ATR']
    df['smaLow'] = df['low'].rolling(length).mean() - df['ATR']
    df['hlv'] = np.where(df['close'] > df['smaHigh'], 1, np.where(df['close'] < df['smaLow'], -1, np.NAN))
    df['hlv'] = df['hlv'].ffill()
    df['sslDown'] = np.where(df['hlv'] < 0, df['smaHigh'], df['smaLow'])
    df['sslUp'] = np.where(df['hlv'] < 0, df['smaLow'], df['smaHigh'])

    return df['sslDown'], df['sslUp']

def funcNadarayaWatsonEnvelope(dtloc, source = 'close', bandwidth = 8, window = 500, mult = 3):
    """
    // This work is licensed under a Attribution-NonCommercial-ShareAlike 4.0 International (CC BY-NC-SA 4.0) https://creativecommons.org/licenses/by-nc-sa/4.0/
    // Nadaraya-Watson Envelope [LUX]
      https://www.tradingview.com/script/Iko0E2kL-Nadaraya-Watson-Envelope-LUX/
     :return: up and down   
     translated for freqtrade: viksal1982  viktors.s@gmail.com  


     df.shape[0]
    """ 
    dtNWE = dtloc.copy()
    dtNWE['nwe_up'] = np.nan
    dtNWE['nwe_down'] =  np.nan
    wn = np.zeros((window, window))
    for i in range(window):
        for j in range(window):
                wn[i,j] = math.exp(-(math.pow(i-j,2)/(bandwidth*bandwidth*2)))
    sumSCW = wn.sum(axis = 1)

    def calc_nwa(dfr, init=0):
        global calc_src_value
        if init == 1:
            calc_src_value = list()
            return
        calc_src_value.append(dfr[source])
        mae = 0.0
        y2_val = 0.0
        y2_val_up = np.nan
        y2_val_down = np.nan
        if len(calc_src_value) > window:
            calc_src_value.pop(0)
        if len(calc_src_value) >= window:
            src = np.array(calc_src_value)
            sumSC = src * wn
            sumSCS = sumSC.sum(axis = 1)
            y2 = sumSCS / sumSCW
            sum_e = np.absolute(src - y2)
            mae = sum_e.sum()/window*mult 
            y2_val = y2[-1]
            y2_val_up = y2_val + mae
            y2_val_down = y2_val - mae
        return y2_val_up,y2_val_down
    calc_nwa(None, init=1)
    dtNWE[['nwe_up','nwe_down']] = dtNWE.apply(calc_nwa, axis = 1, result_type='expand')
    return dtNWE[['nwe_up','nwe_down']]

class NWE(IStrategy):
   
    INTERFACE_VERSION = 2


    window_buy = IntParameter(60, 1000, default=300, space='buy', optimize=True)
    bandwidth_buy = IntParameter(2, 15, default=9, space='buy', optimize=True)
    mult_buy = DecimalParameter(0.5, 20.0, default=4, space='buy', optimize=True)

   # ROI table:
    minimal_roi = {
        "0": 0.103,
        "3": 0.07,
        "5": 0.05,
        "60": 0.025,
        "120": 0.012,
        "240": 0.007,
    }

    # Stoploss:
    stoploss = -0.046  # use custom stoploss

    # Trailing stop:
    trailing_stop = True
    trailing_stop_positive = 0.002
    trailing_stop_positive_offset = 0.012
    trailing_only_offset_is_reached = False

    # Optimal timeframe for the strategy.
    timeframe = '5m'
    custom_info = {}
    custom_trade_info = {}
    # Run "populate_indicators()" only for new candle.
    process_only_new_candles = False

    # These values can be overridden in the "ask_strategy" section in the config.
    use_sell_signal = True
    sell_profit_only = False
    ignore_roi_if_buy_signal = False

    # Number of candles the strategy requires before producing valid signals
    startup_candle_count: int = 300

    # Optional order type mapping.
    order_types = {
        'buy': 'limit',
        'sell': 'limit',
        'stoploss': 'market',
        'stoploss_on_exchange': False
    }

    # Optional order time in force.
    order_time_in_force = {
        'buy': 'gtc',
        'sell': 'gtc'
    }
    
    plot_config = {
        # Main plot indicators (Moving averages, ...)
        'main_plot': {
            'nwe_up': {'color': 'red'},
            'nwe_down': {'color': "rgba(155,150,200,2.4)"},
            'ema_100': {'color': 'blue'},
            'ema_90': {'color': 'green'},
        },
        'subplots': {

          
        }
    }

    def informative_pairs(self):

        return []

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        if not metadata['pair'] in self.custom_trade_info:
            self.custom_trade_info[metadata['pair']] = {}

        if not 'count' in self.custom_trade_info[metadata["pair"]]:
            self.custom_trade_info[metadata['pair']]['count'] = 0
        
        if not 'forsell' in self.custom_trade_info[metadata["pair"]]:
            self.custom_trade_info[metadata['pair']]['forsell'] = 0


        dataframe['ema_100'] = ta.EMA(dataframe, timeperiod=100)
        dataframe['ema_90'] = ta.EMA(dataframe, timeperiod=90)
        dataframe[['nwe_up','nwe_down']] = funcNadarayaWatsonEnvelope(dataframe, source = 'close', bandwidth = self.bandwidth_buy.value, window = self.window_buy.value, mult = self.mult_buy.value)
        return dataframe

    def populate_buy_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        conditions = []

        con1 = ( 
                 (qtpylib.crossed_below(dataframe['close'], dataframe['nwe_down'])) & 
                 (dataframe['ema_100'] > dataframe['close'] ) &
                 (dataframe['ema_90'] > dataframe['nwe_up'] ) &
                 (dataframe['volume'] > 0)  # Make sure Volume is not 0
            )

        con2 = (
              
                (self.custom_trade_info[metadata['pair']]['count'] == 1) & 
                (dataframe['volume'] > 0)
            )



        conditions.append(con1)
        conditions.append(con2)
        # conditions.append(con3)
        
        dataframe.loc[con1, 'buy_tag'] = " con1 "
        dataframe.loc[con2, 'buy_tag'] = " continue  "
        # dataframe.loc[con3, 'buy_tag'] = " con3 "

        if conditions:
            dataframe.loc[
                reduce(lambda x, y: x | y, conditions),
                'buy'
            ] = 1

        return dataframe
    
    def confirm_trade_entry(self, pair: str, order_type: str, amount: float, rate: float, time_in_force: str, **kwargs) -> bool:
                
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        last_candle = dataframe.iloc[-1].squeeze()
        last_candle2 = dataframe.iloc[-2].squeeze()

        logger.info("dentro de entry")

        if (last_candle['close'] < last_candle2['close']) :
            logger.info("comprobando closes")
            self.custom_trade_info[pair]['count'] = 1
            return False
        else :
            self.custom_trade_info[pair]['count'] = 0
            logger.info("suma y sigue ")            
            return True
   
    def populate_sell_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        
        conditions = []

        con1 = ( 
                 (dataframe['close'] > dataframe['nwe_up'])
            )

        con2 = (
              
                (self.custom_trade_info[metadata['pair']]['forsell'] == 1)
            )

        conditions.append(con1)
        conditions.append(con2)       

        if conditions:
            dataframe.loc[
                reduce(lambda x, y: x | y, conditions),
                'sell'
            ] = 1

        return dataframe
        
    def confirm_trade_exit(self, pair: str, trade: Trade, order_type: str, amount: float,rate: float, time_in_force: str, sell_reason: str,current_time: datetime, **kwargs) -> bool:

        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)

        last_candle = dataframe.iloc[-1]
        previous_candle_1 = dataframe.iloc[-2]

        logger.info("dentro de sell")
        if (self.custom_trade_info[pair]['forsell'] == 1) :
            if (last_candle['close'] >= previous_candle_1['close']) :
                self.custom_trade_info[pair]['forsell'] = 1
                logger.info("no vende")
                return None
            else:
                self.custom_trade_info[pair]['forsell'] = 0
                logger.info(f"ponemos a cero {pair}")
                return "sell_by_nwe"
        else:
            self.custom_trade_info[pair]['forsell'] = 1
            logger.info("no vende esta a cero")
            return None

```

```

**Comment by xmatthias** [comment]
The problem is your usage of the variables, not the variables themselfs.

The [bot execution logic](https://www.freqtrade.io/en/stable/bot-basics/#bot-execution-logic) documentation highlights the differences between the two modes.

`populate_*`  methods will be executed only once during backtesting (at the beginning), while callbacks will be executed thereafter, while simulating trades.
As such, prior trades are not available in any `populate_*` method - and neither are variables that change as part of callbacks.

You can probably build most of the logic here with different callbacks (like custom_sell instead of the forcesell variable). It'll simply require a bit different thinking - but the way it's now, it'll clearly not work for backtesting.

**Comment by anakein** [comment]
ok thanks

**Comment by anakein** [comment]
i have one problem with this strategy in live/dry run.
if I forcesell by telegram or frequi, don´t sell.

How can confirm_trade_exit function know that it is a forcesell?

**Comment by xmatthias** [comment]
You get the exit reason as parameter - just verify that - in new versions (2022.4+) you can use `"force_exit"`  as check for force exits.
